protocol/morse/context.go

```diff
@@ -209,3 +165,86 @@ func (rc *requestContext) sendRelay(
+// - Tracks the endpoint success in observations
+// - Builds and returns the protocol response from the endpoint's returned data.
+func (rc *requestContext) handleEndpointSuccess(
+	endpointQueryTime time.Time,
+	endpointResponsePayload []byte,
+	endpointResponseHTTPStatusCode int,
+) (protocol.Response, error) {
+	// Track the endpoint success observation.
+	rc.endpointObservations = append(rc.endpointObservations,
+		buildEndpointSuccessObservation(
+			*rc.selectedEndpoint,
+			endpointQueryTime,
+			time.Now(), // Timestamp: endpoint query completed.
+		),
+	)
+
+	// Build and return the relay response received from the endpoint.
+	return protocol.Response{
+		EndpointAddr:   rc.selectedEndpoint.Addr(),
+		Bytes:          endpointResponsePayload,
+		HTTPStatusCode: endpointResponseHTTPStatusCode,
+	}, nil
+}
+
+// handleEndpointError records an endpoint error observation and returns the response.
+// - Tracks the endpoint error in observations
+// - Builds and returns the protocol response from the endpoint's returned data.
+func (rc *requestContext) handleEndpointError(
+	endpointQueryTime time.Time,
+	endpointErr error,
+	endpointResponsePayload []byte,
+	endpointResponseHTTPStatusCode int,
+) (protocol.Response, error) {
+	// Classify the endpoint's error for the observation.
+	// Determine any applicable sanctions.
+	endpointErrorType, recommendedSanctionType := classifyRelayError(rc.logger, endpointErr)
+
+	// Log the endpoint error.
+	hydratedLogger := rc.getHydratedLogger("HandleServiceRequest")
+	hydratedLogger.Error().
+		Str("error_type", endpointErrorType.String()).
+		Str("sanction_type", recommendedSanctionType.String()).
+		Err(endpointErr).
+		Msg("relay error occurred.")
+
+	// Track the endpoint error observation.
+	rc.endpointObservations = append(rc.endpointObservations,
+		buildEndpointErrorObservation(
+			*rc.selectedEndpoint,
+			endpointQueryTime,
+			time.Now(), // Timestamp: endpoint query completed.
+			endpointErrorType,
+			fmt.Sprintf("relay error: %v", endpointErr),
+			recommendedSanctionType,
+		),
+	)
+
+	// Return the endpoint's response as-is, along with the encountered error.
+	return protocol.Response{
+		EndpointAddr:   rc.selectedEndpoint.Addr(),
+		Bytes:          endpointResponsePayload,
+		HTTPStatusCode: endpointResponseHTTPStatusCode,
+	}, endpointErr
+}
+
+// handleInternalError is called if the request processing fails.
+// This only happens before the request is sent to any endpoints.
+// DEV_NOTE: This should NEVER happen and any logged entries by this method should be investigated.
+// - Records an internal error on the request for observations.
+// - Logs an error entry.
+func (rc *requestContext) handleInternalError(internalErr error) (protocol.Response, error) {
+	hydratedLogger := rc.getHydratedLogger("handleInternalError")
+
+	// log the internal error.
+	hydratedLogger.Error().Err(internalErr).Msg("Internal error occurred. This should be investigated as a bug.")
```

Comment:
Given the context can we make this log stand out a bit more? Generally I'm against all-caps logs but would make an exception (or maybe even an emoji 😉).

Reply:
Suggestion: let's use error level of the log for signaling importance.
This is logged at error level: we will minimize such entries and make sure to investigate any error-level logs.
Please reopen if this needs to be revisited.